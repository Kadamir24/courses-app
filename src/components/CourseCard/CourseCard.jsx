import React from 'react';
import styled from 'styled-components';
import { Button } from '../Button/Button';
import { timeConverter } from '../../utils/functions';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import { actionCreators } from '../../store/courses/actionCreators';
import store from '../../store/index';
import { fetchDelete } from '../../utils/api';

const CardContainer = styled.div`
	width: 95%;
	margin: 20px auto;
	border: 2px solid red;
	display: flex;
	justify-content: space-around;
`;

const MainInfo = styled.div`
	width: 60%;
`;

const SubInfo = styled.div`
	width: 30%;
`;

const CourseCard = ({
	title,
	description,
	authors,
	duration,
	creationDate,
	authorsList,
	id,
}) => {
	const token = localStorage.getItem('token');
	const dispatch = useDispatch();
	const deleteCourse = () => {
		dispatch(actionCreators.deleteCourse(id));
		console.log('Check store2', store.getState());
		fetchDelete('courses', id, token);
	};
	return (
		<CardContainer>
			<MainInfo>
				<h2>{title}</h2>
				<div>{description}</div>
			</MainInfo>
			<SubInfo>
				<div>
					Authors:{' '}
					{authorsList.map((author) => {
						let res = authors.includes(author.id) ? author.name + ', ' : '';
						return res;
					})}
				</div>
				<div>Duration: {timeConverter(duration)} hours</div>
				<div>Created: {creationDate} hours</div>
				<Link to={`/courses/${id}`}>
					<Button>Show course</Button>
				</Link>
				<Button>Update</Button>
				<Button onClick={deleteCourse}>Delete course</Button>
			</SubInfo>
		</CardContainer>
	);
};

CourseCard.propTypes = {
	title: PropTypes.string.isRequired,
	description: PropTypes.string.isRequired,
	authors: PropTypes.array.isRequired,
	duration: PropTypes.number.isRequired,
	creationDate: PropTypes.string.isRequired,
	authorsList: PropTypes.array.isRequired,
	id: PropTypes.string.isRequired,
};

export default CourseCard;
