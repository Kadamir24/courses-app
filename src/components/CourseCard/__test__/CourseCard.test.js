import React from 'react';
import ReactDOM from 'react-dom';
import CourseCard from '../CourseCard';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';

import { render } from '@testing-library/react';
const middlewares = [];
const mockStore = configureStore(middlewares);

describe('CourseCard testing', () => {
	it('renders without crashing', () => {
		const initialState = {
			authentication: {
				token: 'Bearer 123123123',
			},
		};
		const store = mockStore(initialState);
		const div = document.createElement('div');
		ReactDOM.render(
			<Provider store={store}>
				<MemoryRouter>
					<CourseCard
						title={'JavaScript'}
						description={'Lorem'}
						authors={['123123', '123123']}
						duration={500}
						creationDate={'01/01/01'}
						authorsList={['123', '123']}
						id={'123124'}
					/>
				</MemoryRouter>
			</Provider>,
			div
		);
	});

	it('renders without crashing', () => {
		const initialState = {
			authentication: {
				token: 'Bearer 123123123',
			},
		};
		const store = mockStore(initialState);
		const { getByTestId, asFragment } = render(
			<Provider store={store}>
				<MemoryRouter>
					<CourseCard
						title={'JavaScript'}
						description={'Lorem'}
						authors={[
							'9b87e8b8-6ba5-40fc-a439-c4e30a373d36',
							'1c972c52-3198-4098-b6f7-799b45903199',
						]}
						duration={500}
						creationDate={'01/01/01'}
						authorsList={[
							{ name: 'author', id: '9b87e8b8-6ba5-40fc-a439-c4e30a373d36' },
							{ name: 'author2', id: '1c972c52-3198-4098-b6f7-799b45903199' },
						]}
						id={'123124'}
					/>
				</MemoryRouter>
			</Provider>
		);
		const firstRender = asFragment();
		expect(getByTestId('title')).toHaveTextContent('JavaScript');
		expect(getByTestId('description')).toHaveTextContent('Lorem');
		expect(getByTestId('duration')).toHaveTextContent('Duration: 08:20 hours');
		expect(getByTestId('authors')).toHaveTextContent(
			'Authors: author, author2,'
		);
		expect(getByTestId('date-creation')).toHaveTextContent(
			'Created: 01/01/01 hours'
		);
		expect(firstRender).toMatchSnapshot(asFragment());
	});
});
