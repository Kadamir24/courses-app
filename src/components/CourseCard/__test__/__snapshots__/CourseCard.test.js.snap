// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CourseCard testing renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="sc-gtsrHT euWWUH course-card"
  >
    <div
      class="sc-dlnjwi fRFoqb"
    >
      <h2
        data-testid="title"
      >
        JavaScript
      </h2>
      <div
        data-testid="description"
      >
        Lorem
      </div>
    </div>
    <div
      class="sc-hKFxyN jRahdW"
    >
      <div
        data-testid="authors"
      >
        Authors: author, author2, 
      </div>
      <div
        data-testid="duration"
      >
        Duration: 08:20 hours
      </div>
      <div
        data-testid="date-creation"
      >
        Created: 01/01/01 hours
      </div>
      <a
        href="/courses/123124"
      >
        <button
          class="sc-bdnxRM CXauk"
        >
          Show course
        </button>
      </a>
    </div>
  </div>
</DocumentFragment>
`;
